### ISSUE TITLE
Bracketed matches an empty body even when its content grammar is required

### ISSUE THREAD

**Issue by cotsupa** [sqlfluff/sqlfluff#8368]
### Search before asking

- [x] I searched the issues and found no similar issues.

### What Happened

`Bracketed(...)` successfully matches empty brackets `()` even when its content grammar contains required (non-optional) elements. In `Bracketed.match` (`src/sqlfluff/core/parser/grammar/sequence.py`), the content match result is never checked: when the content grammar fails, matching proceeds directly to the end-bracket check, so `()` (where the end bracket immediately follows the start bracket) always matches, in STRICT parse mode too. Non-empty invalid content is correctly rejected (the end-bracket match fails at the leftover content), so this only affects the empty-body case.

This makes it impossible for a dialect to express "these brackets must not be empty". Examples in the Snowflake dialect: `WORKLOAD_IDENTITY = ()` (docs require `TYPE`), `OIDC_AUDIENCE_LIST = ()` (requires at least one string), `IN ()` — all parse without an unparsable section today. Spotted while working on #8367 (see review discussion there).

### Expected Behaviour

When the content grammar of a `Bracketed` cannot match an empty sequence (i.e. it contains required elements), empty brackets should fail to match in STRICT mode (or produce an unparsable section in GREEDY mode).

### Observed Behaviour

Empty brackets always match regardless of the content grammar.

### How to reproduce

```python
from sqlfluff.core import Linter

linter = Linter(dialect="snowflake")
parsed = linter.parse_string("ALTER USER u SET WORKLOAD_IDENTITY = ();")
print([v for v in parsed.violations if v.rule_code() == "PRS"])  # [] — no parse error
```

### Dialect

snowflake (but the behaviour is dialect-independent — it's in the core `Bracketed` grammar)

### Version

main (f26ed33), also reproduces on 4.3.0

### Configuration

Default.

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by mameikagou** [comment]
Thanks for the clear reproduction. I verified that this still affects current `main` and that there is no linked fix. Since you checked that you are willing to submit a PR, are you still planning to work on it? If not, I would be happy to prepare the focused core-parser fix and regression tests; I do not want to duplicate or preempt your work.

**Comment by cotsupa** [comment]
@mameikagou Thanks for checking in, and for the offer — much appreciated. Yes, I'm still planning to work on this and will take it. I'll open a PR shortly.
